walk/walk.go

```diff
@@ -154,63 +155,48 @@ func visit(c *config.Config, cexts []config.Configurer, knownDirectives map[stri
 		haveError = true
 	}
 
-	collectionOnly := f == nil && getWalkConfig(c).updateOnly
+	c.ValidBuildFileNames = trie.walkConfig.validBuildFileNames
 
 	// Configure the current directory if not only collecting files
-	if !collectionOnly {
-		configure(cexts, knownDirectives, c, rel, f)
-	}
+	configure(cexts, knownDirectives, c, trie.rel, f)
 
-	wc := getWalkConfig(c)
-	if wc.isExcluded(rel) {
-		return nil, false
-	}
+	wc := trie.walkConfig
 
 	// Filter and collect files
-	var regularFiles []string
-	for _, ent := range trie.files {
-		base := ent.Name()
-		entRel := path.Join(rel, base)
-		if wc.isExcluded(entRel) {
-			continue
-		}
-		if ent := resolveFileInfo(wc, dir, entRel, ent); ent != nil {
-			regularFiles = append(regularFiles, base)
-		}
+	regularFiles := make([]string, 0, len(trie.files))
+	for entSub := range trie.files {
+		regularFiles = append(regularFiles, entSub)
 	}
 
-	shouldUpdate := updateRels.shouldUpdate(rel, updateParent)
+	shouldUpdate := updateRels.shouldUpdate(trie.rel, updateParent)
+
+	// Ensure a deterministic order for regular files and subdirectories.
+	// This includes both the params to the callback as well as the `visit` order.
+	slices.Sort(regularFiles)
+
+	// TODO: don't sort content of trie, pre-sort it or sort only a local copy
```

Comment:
IMO both of the `.Sort*` calls being here isn't ideal. It would be nicer if the `trie` struct already had all the nodes sorted and readonly, but the need for a `map` to store the path makes that more difficult atm. Maybe another thing that can be cleaned up later?